aws/warehouse.go

```diff
@@ -80,6 +88,96 @@ type AccountWarehouse struct {
 	svcUserArn *string
 	store      storage.Store
 	tmp        map[string]iam.AccessKey
+	keyGC      *processaws.KeyGC
+}
+
+func (wh *AccountWarehouse) GetServiceAccounts(ctx context.Context, project string) (<-chan *clouds.Account, error) {
+	c := make(chan *clouds.Account)
+	go func() {
+		defer close(c)
+		f := func(acct *iam.User) error {
+			a := &clouds.Account{
+				ID:          aws.StringValue(acct.UserId),
+				DisplayName: aws.StringValue(acct.UserName),
+			}
+			select {
+			case c <- a:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+			return nil
+		}
+		sess, err := createSession()
+		if err !=nil {
+			return
+		}
+		svc := iam.New(sess)
+		// FIXME: get PathPrefix from config
+		accounts, err := svc.ListUsers(&iam.ListUsersInput{
+			PathPrefix: aws.String("/ddap/"),
+		})
+		users := accounts.Users
+		for _, user := range users {
+			if err := f(user); err != nil {
+				glog.Errorf("getting user accounts list: %v", err)
+				return
+			}
+		}
+
+	}()
+	return c, nil
+}
+
+func (wh *AccountWarehouse) RemoveServiceAccount(ctx context.Context, project, accountID string) error {
+	// TODO
+	// Unlike the AWS Management Console, when
+	//       you delete a user programmatically, you must delete the items  attached
+	//       to  the user manually, or the deletion fails. For more information, see
+	//       Deleting an IAM User . Before attempting to delete a user,  remove  the
+	//       following items
+	// Refer: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_manage.html#id_users_deleting_cli
+	panic("implement me removeserviceaccount")
+}
+
+//This method is the main method where key removal happens
+func (wh *AccountWarehouse) ManageAccountKeys(ctx context.Context, project, accountID string, ttl, maxKeyTTL time.Duration, now time.Time, keysPerAccount int64) (int, int, error) {
+	expired := now.Add(-1 * maxKeyTTL).Format(time.RFC3339)
+	sess, err := createSession()
+	if err != nil {
+		return 0, 0, fmt.Errorf("error creating session: %v", err)
+	}
+	svc := iam.New(sess)
+	accessKeys, err := svc.ListAccessKeys(&iam.ListAccessKeysInput{
+		UserName: aws.String(accountID),
+	})
+	if err != nil {
+		return 0, 0, fmt.Errorf("error getting key list: %v", err)
+	}
+	keys := accessKeys.AccessKeyMetadata
+	var actives []*iam.AccessKeyMetadata
+	active := len(keys)
+	//keys := accessKeys.AccessKeyMetadata
+	for _, key := range keys {
+		t := timeutil.TimestampProto(aws.TimeValue(key.CreateDate))
+		if timeutil.RFC3339(t) < expired {
+			// Access key deletion
+			_, err := svc.DeleteAccessKey(&iam.DeleteAccessKeyInput{
+				AccessKeyId: key.AccessKeyId,
+				UserName:    aws.String(accountID),
+			})
+			if err !=nil {
+				return active, len(keys) - active, fmt.Errorf("error deleting access key: %v", err)
```

Comment:
If I'm reading this correctly, we bail out of the whole background process if we encounter an error deleting one single key, right? Is this the same logic in the GCP version that we copied?

Reply:
Yeah, it is the same logic as the GCP one. We exit from the loop if there is an error deleting a key and subsequent keys in the array will not be processed.